llmfoundry/callbacks/curriculum_learning_callback.py

```diff
@@ -0,0 +1,103 @@
+
+log = logging.getLogger(__name__)
+
+
+class CurriculumLearning(Callback):
+    """Starts an epoch with a different dataset when resuming from a checkpoint.
+
+    This callback is currently experimental. The API may change in the future.
+
+    Args:
+        dataset_index (int): The index of the dataset currently being used.
+        current_dataset_config (Dict): The configuration of the dataset currently
+            being used.
+        dataset (StreamingDataset): The dataset currently being used.
+    """
+
+    def __init__(self, dataset_index: int, dataloader: Union[DataLoader,
+                                                             Iterable],
+                 current_dataset_config: Dict):
+        self.dataset_index = dataset_index
+        self.saved_dataset_index = 0
+        self.all_dataset_configs = []
+        # The current dataset config is resolved and passed in train.py
+        self.current_dataset_config = current_dataset_config
+        self.new_dataset_setup = False
+        self.dataset_config_appended = False
+
+        # Must pass in dataset directly since it is not actually accessible at Event.INIT in
+        # Composer. We need to get the new dataset state to override checkpoint dataset state.
+        # Check if we are using a DataLoader and StreamingDataset
+        if not isinstance(dataloader, DataLoader):
+            raise ValueError(
+                f'CurriculumLearning callback can only be used with a train ',
+                f'dataloader of type DataLoader, but got {type(dataloader)}.')
+        dataset = dataloader.dataset
+        if not isinstance(dataset, StreamingDataset):
+            raise ValueError(
+                f'CurriculumLearning callback only supports StreamingDataset ',
+                f'because it requires loading and saving dataset state. ',
+                f'Instead, got a dataset of type {type(dataset)}')
+        # Save the current dataset state so we can restore it if needed.
+        self.current_dataset_state = dataset.state_dict(0, False)
+
+    def after_load(self, state: State, logger: Logger):
+        del logger
+
+        # As saved_dataset_index is loaded from state_dict, this only run when
+        # a user explicitly increments the dataset_index and not on any other
+        # resumption, including autoresume.
+        train_loader = state._train_dataloader
+        assert isinstance(
+            train_loader,
+            DataLoader), 'CurriculumLearning callback requires a DataLoader.'
+        dataset = train_loader.dataset
+        assert isinstance(
+            dataset, StreamingDataset
+        ), 'CurriculumLearning callback requires a StreamingDataset.'
+        if self.saved_dataset_index < self.dataset_index:
+            # Ignore the dataset state that was read in from the checkpoint, and
+            # replace with the new dataset state. This preserves resumption info.
+            if self.current_dataset_state['epoch'] < 0:
```

Comment:
How can epoch be < 0?

Reply:
When a streamingdataset is first initialized and we call `state_dict`, the epoch that gets returned is `-1` because we increment the epoch counter at the start of  `__iter__`, which has not yet been called. So to set the state dict correctly, we have to check if `epoch` is -1.

Reply:
Can you extend comment to explain this scenario?

Reply:
added!